p2panda-rs/src/schema/schema.rs

```diff
@@ -253,43 +196,95 @@ impl Schema {
     /// It has the format "<schema name>__<hashed schema document view>" for application schemas
     /// and "<schema_name>__<version>" for system schemas (note that this has two underscores,
     /// while schema id has only one).
-    #[allow(unused)]
     pub fn hash_id(&self) -> String {
         match self.id.version() {
             SchemaVersion::Application(view_id) => {
                 format!("{}__{}", self.name(), DocumentViewHash::from(&view_id))
             }
             SchemaVersion::System(version) => {
-                format!("{}__{}", self.name(), &version.to_string())
+                format!("{}__{}", self.name(), version)
             }
         }
     }
 
     /// Access the schema version.
-    #[allow(unused)]
     pub fn version(&self) -> SchemaVersion {
         self.id.version()
     }
 
     /// Access the schema name.
-    #[allow(unused)]
     pub fn name(&self) -> &str {
         self.id.name()
     }
 
     /// Access the schema description.
-    #[allow(unused)]
     pub fn description(&self) -> &str {
         &self.description
     }
 
     /// Access the schema fields.
-    #[allow(unused)]
-    pub fn fields(&self) -> &BTreeMap<FieldKey, FieldType> {
+    pub fn fields(&self) -> &BTreeMap<FieldName, FieldType> {
         &self.fields
     }
 }
 
+#[cfg(any(feature = "testing", test))]
+impl Schema {
+    /// Create an application schema instance with the given id, description and fields.
+    ///
+    /// Use [`Schema::get_system`] to access static system schema instances.
+    ///
+    /// ## Example
+    ///
+    /// ```
+    /// # #[cfg(test)]
+    /// # mod doc_test {
+    /// # extern crate p2panda_rs;
+    /// # use p2panda_rs::document::DocumentViewId;
+    /// # use p2panda_rs::test_utils::fixtures::{document_view_id};
+    /// #
+    /// # #[rstest]
+    /// # fn main(#[from(document_view_id)] schema_document_view_id: DocumentViewId) {
+    /// let schema = Schema::new(
+    ///     SchemaId::Application("cucumber", schema_document_view_id),
+    ///     "A variety in the cucumber society's database.",
+    ///     vec![
+    ///         ("name", FieldType::String),
+    ///         ("grow_cycle_days", FieldType::Int),
+    ///         ("flavor_rating", FieldType::Int),
+    ///     ]
+    /// );
+    /// assert!(schema.is_ok());
+    /// # }
+    /// # }
+    /// ```
+    pub fn new(
```

Comment:
I'd be up for making this public to be able to create schema instances without a persistence and materialization layer. I'm afraid that there is otherwise quite a bunch to do before you can write an operation. What do you think?

Reply:
Yeh, I'd be up for that.

Reply:
Okay, I think I can add it using the new validation methods :+1: